### PR TITLE
Goa 1999 filter annotation by GO id

### PR DESCRIPTION
**Pull request by TonyWardell**
As a QuickGO user
I would like to be able to filter annotations using an GO ontology id(s)
So that I can see annotations that are associated to that/those particular ontology(s)
Acceptance criteria:
Unknown ontology id should show an error with an explicit message
If the ontology id does not match any of the ids on the annotation list, then the result should be empty
If the ontology id matches the ids of ontologies in the list of annotations, then the result should only show the affected annotations
The user can input more than one ontology id, when this is the case then the ids should be ORed.
This filter criteria can be used in conjunction with other non-related filter criteria
The filter should be case-insensitive

### CODE REVIEW

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java [hunk 1]
```diff
@@ -352,6 +354,60 @@ public void failToFindAnnotationsWhenQualifierDoesntExist() throws Exception {
 
     }
 
+    //---------- Gene Ontology Id
+
+    @Test
+    public void successfullyLookupAnnotationsByGoId() throws Exception {
+        String goId = "GO:0003824";
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: It'd be clearer if this was a constant with a descriptive name

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java [hunk 2]
```diff
@@ -352,6 +354,60 @@ public void failToFindAnnotationsWhenQualifierDoesntExist() throws Exception {
...
+                .andExpect(contentTypeToBeJson())
+                .andExpect(totalNumOfResults(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(fieldsInAllResultsExist(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(itemExistsExpectedTimes(GO_ID, goId,3));
+    }
+
+    @Test
+    public void successfullyLookupAnnotationsByGoIdCaseInsensitive() throws Exception {
+        String argumentFromClient = "go:0003824";
+        String goId = "GO:0003824";
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: It'd be clearer if this was a constant with a descriptive name

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java [hunk 3]
```diff
@@ -352,6 +354,60 @@ public void failToFindAnnotationsWhenQualifierDoesntExist() throws Exception {
...
+                .andExpect(status().isOk())
+                .andExpect(contentTypeToBeJson())
+                .andExpect(totalNumOfResults(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(fieldsInAllResultsExist(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(itemExistsExpectedTimes(GO_ID, goId,3));
+    }
+
+    @Test
+    public void successfullyLookupAnnotationsByGoIdCaseInsensitive() throws Exception {
+        String argumentFromClient = "go:0003824";
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: could refer to a constant `String` that calls `.toLowercase()`

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestValidationIT.java [hunk 4]
```diff
@@ -230,6 +230,59 @@ public void oneValidTaxIdAndOneInvalidTaxIdIsInvalid() {
...
+    @Test
+    public void mixedCaseGoIdIsValid() {
+        String[] goIds = {"GO:0003824", "gO:0003824", "Go:0003824"};
+
+        Arrays.stream(goIds).forEach(
+                mixedCaseId -> {
+                    annotationRequest.setGoId(mixedCaseId);
+
+                    Set<ConstraintViolation<AnnotationRequest>> violations = validator.validate(annotationRequest);
+                    printConstraintViolations(violations);
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestValidationIT.java]: necessary?

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java [hunk 5]
```diff
@@ -352,6 +354,60 @@ public void failToFindAnnotationsWhenQualifierDoesntExist() throws Exception {
...
+                .andExpect(status().isOk())
+                .andExpect(contentTypeToBeJson())
+                .andExpect(totalNumOfResults(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(fieldsInAllResultsExist(NUMBER_OF_GENERIC_DOCS))
+                .andExpect(itemExistsExpectedTimes(GO_ID, goId,3));
+    }
+
+    @Test
+    public void failToFindAnnotationsWhenGoIdDoesntExist() throws Exception {
+        String goId = "GO:0009871";
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: It'd be clearer if this was a constant with a descriptive name

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java [hunk 6]
```diff
@@ -352,6 +354,60 @@ public void failToFindAnnotationsWhenQualifierDoesntExist() throws Exception {
...
+        response.andDo(print())
+                .andExpect(status().isOk())
+                .andExpect(contentTypeToBeJson())
+                .andExpect(totalNumOfResults(0));
+    }
+
+
+    @Test
+    public void incorrectFormattedGoIdCausesError() throws Exception {
+        String goId = "GO:1";
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: It'd be clearer if this was a constant with a descriptive name

File: annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java [hunk 7]
```diff
@@ -58,7 +58,14 @@ public static boolean isRetrievable(String field) {
         public static final String GO_EVIDENCE = storeAndGet(VALUES, AnnotationFields.GO_EVIDENCE);
         public static final String QUALIFIER   = storeAndGet(VALUES, AnnotationFields.QUALIFIER);
         public static final String REFERENCE_SEARCH = storeAndGet(VALUES, AnnotationFields.REFERENCE_SEARCH);
+        static {
```
Comment (eddturner) [inline annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java]: What's the intent behind this static block?

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java (hunk 1 above)
Comment (TonyWardell) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: Done.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java (hunk 3 above)
Comment (TonyWardell) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: Done.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestValidationIT.java (hunk 4 above)
Comment (TonyWardell) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestValidationIT.java]: Removed.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java (hunk 5 above)
Comment (TonyWardell) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerIT.java]: Done.

File: annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java (hunk 7 above)
Comment (TonyWardell) [inline annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java]: Make the code clearer and simpler for the storeAndGet operation.
I've now completed the refactor.

Comment (eddturner) [inline annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java]: > "Make the code clearer and simpler for the storeAndGet operation."

How so?

Looking at line 55

```java
// this bit stores AnnotationFields.ASSIGNED_BY in VALUES, and assigns this value to AnnotationFields.Searchable.ASSIGNED_BY
public static final String ASSIGNED_BY = storeAndGet(VALUES, AnnotationFields.ASSIGNED_BY);
```

... and so, the static block seems to do nothing.
```java
static {
            storeAndGet(VALUES, AnnotationFields.ASSIGNED_BY);
...
}
```

Comment (eddturner) [inline annotation-common/src/main/java/uk/ac/ebi/quickgo/annotation/common/document/AnnotationFields.java]: There are also duplicates in the static block, e.g., line 63 and line 69

**Comment by eddturner** [comment]
This branch no longer builds:

```txt
Failed tests: 
  AnnotationRequestValidationIT.goIdIsInvalid:305->lambda$goIdIsInvalid$8:313 
Expected: is "At least one invalid 'GO Id' value is invalid: GO:4"
     but: was "At least one 'GO Id' value is invalid: GO:4"
```

**Pull request merged**